### PR TITLE
[X] compile gridlength and row|defcollection

### LINKED ISSUE #5430: XamlC should compile RowDefinition and ColumnDefinition CollectionTypeConverters
### Description

After compiling the .NET Podcast app with Release, I'm seeing a lot of code in InitializeComponents that look like:

```C#
grid4.SetValue(Grid.RowDefinitionsProperty, new RowDefinitionCollectionTypeConverter().ConvertFromInvariantString("56, auto"));

grid.SetValue(Grid.RowDefinitionsProperty, new RowDefinitionCollectionTypeConverter().ConvertFromInvariantString("*, auto"));

grid.SetValue(Grid.RowDefinitionsProperty, new RowDefinitionCollectionTypeConverter().ConvertFromInvariantString("auto,auto,auto,auto,auto,auto"));

```

Similar to https://github.com/dotnet/maui/issues/5018 we should parse and convert these strings at compile time.

### Steps to Reproduce

1. `dotnet build Microsoft.NetConf2021.Maui.csproj -t:Run -c Release -f net6.0-android -r android-arm64` on the https://github.com/microsoft/dotnet-podcasts app
2. Inspect the resulting IL for calls to ConvertFromInvariantString

### Version with bug

Unknown/Other (please specify)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS

### Affected platform versions

All

### Did you find any workaround?

_No response_

### Relevant log output

```shell
cc @StephaneDelcroix
```

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
- fixes #5430

### Issues Fixed

Fixes #

### CODE REVIEW

File: src/Controls/src/Core/ColumnDefinition.cs [hunk 1]
```diff
@@ -6,15 +6,17 @@ namespace Microsoft.Maui.Controls
...
+		public static readonly BindableProperty WidthProperty = BindableProperty.Create(nameof(Width), typeof(GridLength), typeof(ColumnDefinition), GridLength.Star,
 			propertyChanged: (bindable, oldValue, newValue) => ((ColumnDefinition)bindable).OnSizeChanged());
 
 		/// <include file="../../docs/Microsoft.Maui.Controls/ColumnDefinition.xml" path="//Member[@MemberName='.ctor']/Docs" />
 		public ColumnDefinition()
 		{
-			MinimumWidth = -1;
 		}
 
+		public ColumnDefinition(GridLength width)
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/ColumnDefinition.cs:17]: new API, can be hidden from Intelissense

File: src/Controls/src/Core/ColumnDefinitionCollection.cs [hunk 2]
```diff
@@ -3,5 +3,12 @@ namespace Microsoft.Maui.Controls
 	/// <include file="../../docs/Microsoft.Maui.Controls/ColumnDefinitionCollection.xml" path="Type[@FullName='Microsoft.Maui.Controls.ColumnDefinitionCollection']/Docs" />
 	public sealed class ColumnDefinitionCollection : DefinitionCollection<ColumnDefinition>
 	{
+	    public ColumnDefinitionCollection() : base()
+	    {
+	    }
+
+	    public ColumnDefinitionCollection(params ColumnDefinition[] definitions) : base (definitions)
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/ColumnDefinitionCollection.cs:10]: new API, can be hidden from Intelissense

File: src/Controls/src/Core/RowDefinition.cs [hunk 3]
```diff
@@ -6,13 +6,17 @@ namespace Microsoft.Maui.Controls
...
+		public static readonly BindableProperty HeightProperty = BindableProperty.Create(nameof(Height), typeof(GridLength), typeof(RowDefinition), GridLength.Star,
 			propertyChanged: (bindable, oldValue, newValue) => ((RowDefinition)bindable).OnSizeChanged());
 
 		/// <include file="../../docs/Microsoft.Maui.Controls/RowDefinition.xml" path="//Member[@MemberName='.ctor']/Docs" />
 		public RowDefinition()
 		{
-			MinimumHeight = -1;
+		}
+
+		public RowDefinition(GridLength height)
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/RowDefinition.cs:17]: new API, can be hidden from Intelissense

File: src/Controls/src/Core/RowDefinitionCollection.cs [hunk 4]
```diff
@@ -3,5 +3,12 @@ namespace Microsoft.Maui.Controls
 	/// <include file="../../docs/Microsoft.Maui.Controls/RowDefinitionCollection.xml" path="Type[@FullName='Microsoft.Maui.Controls.RowDefinitionCollection']/Docs" />
 	public sealed class RowDefinitionCollection : DefinitionCollection<RowDefinition>
 	{
+	    public RowDefinitionCollection() : base()
+	    {
+	    }
+
+	    public RowDefinitionCollection(params RowDefinition[] definitions) : base (definitions)
```
Comment (StephaneDelcroix) [inline src/Controls/src/Core/RowDefinitionCollection.cs:10]: new API, can be hidden from Intelissense

File: src/Core/src/Primitives/GridLength.cs [hunk 5]
```diff
@@ -9,22 +9,16 @@ namespace Microsoft.Maui
 	public struct GridLength
 	{
 		/// <include file="../../docs/Microsoft.Maui/GridLength.xml" path="//Member[@MemberName='Auto']/Docs" />
-		public static GridLength Auto
-		{
-			get { return new GridLength(1, GridUnitType.Auto); }
-		}
+		public static readonly GridLength Auto = new GridLength(1, GridUnitType.Auto);
```
Comment (StephaneDelcroix) [inline src/Core/src/Primitives/GridLength.cs:12]: Actual breaking change, from static property to static field, require recompile

File: src/Core/src/Primitives/GridLength.cs [hunk 6]
```diff
@@ -9,22 +9,16 @@ namespace Microsoft.Maui
...
 
 		/// <include file="../../docs/Microsoft.Maui/GridLength.xml" path="//Member[@MemberName='Star']/Docs" />
-		public static GridLength Star
-		{
-			get { return new GridLength(1, GridUnitType.Star); }
-		}
+		public static readonly GridLength Star = new GridLength(1, GridUnitType.Star);
```
Comment (StephaneDelcroix) [inline src/Core/src/Primitives/GridLength.cs:15]: Actual breaking change, from static property to static field, require recompile

File: src/Core/src/Primitives/GridLength.cs [hunk 7]
```diff
@@ -9,22 +9,16 @@ namespace Microsoft.Maui
...
 
 		/// <include file="../../docs/Microsoft.Maui/GridLength.xml" path="//Member[@MemberName='Value']/Docs" />
-		public double Value { get; }
+		public readonly double Value { get; }
```
Comment (StephaneDelcroix) [inline src/Core/src/Primitives/GridLength.cs:18]: this should be fine

File: src/Core/src/Primitives/GridLength.cs [hunk 8]
```diff
@@ -9,22 +9,16 @@ namespace Microsoft.Maui
...
 
 		/// <include file="../../docs/Microsoft.Maui/GridLength.xml" path="//Member[@MemberName='GridUnitType']/Docs" />
-		public GridUnitType GridUnitType { get; }
+		public readonly GridUnitType GridUnitType { get; }
```
Comment (StephaneDelcroix) [inline src/Core/src/Primitives/GridLength.cs:21]: this should be fine

**Review by StephaneDelcroix — COMMENTED** [review]
added comments on API changes

**Comment by Redth** [comment]
Seems ok to me, mostly additive and recompiles but not renames of existing 👍

File: src/Core/src/Primitives/GridLength.cs [hunk 9]
```diff
@@ -9,22 +9,16 @@ namespace Microsoft.Maui
 	public struct GridLength
```
Comment (eerhardt) [inline src/Core/src/Primitives/GridLength.cs:9]: While we are in here making readonly changes, can we just make the whole struct readonly? That is our preferred approach in dotnet/runtime.

```suggestion
	public readonly struct GridLength
```

File: src/Controls/tests/Xaml.UnitTests/DefinitionCollectionTests.xaml.cs [hunk 10]
```diff
@@ -42,6 +45,26 @@ public void DefinitionCollectionsParsedFromMarkup([Values(false, true)] bool use
 				Assert.That(rowdef[4].Height, Is.EqualTo(new GridLength(20, GridUnitType.Absolute)));
 
 			}
+
+			[Test]
+			public void DefinitionCollectionsReplacedAtCompilation()
+			{
+				MockCompiler.Compile(typeof(DefinitionCollectionTests), out var methodDef);
+				Assert.That(!methodDef.Body.Instructions.Any(instr=>InstructionIsDefColConvCtor(methodDef, instr)), "This Xaml still genereate [Row|Col]efinitionConverter ctor");
```
Comment (eerhardt) [inline src/Controls/tests/Xaml.UnitTests/DefinitionCollectionTests.xaml.cs:53]:
```suggestion
				Assert.That(!methodDef.Body.Instructions.Any(instr=>InstructionIsDefColConvCtor(methodDef, instr)), "This Xaml still generates [Row|Col]DefinitionCollectionTypeConverter ctor");
```

File: src/Controls/src/Core/RowDefinitionCollectionTypeConverter.cs [hunk 11]
```diff
@@ -24,11 +26,8 @@ public override object ConvertFrom(ITypeDescriptorContext context, CultureInfo c
 			if (strValue != null)
 			{
 				var lengths = strValue.Split(',');
-				var coldefs = new RowDefinitionCollection();
 				var converter = new GridLengthTypeConverter();
-				foreach (var length in lengths)
-					coldefs.Add(new RowDefinition { Height = (GridLength)converter.ConvertFromInvariantString(length) });
-				return coldefs;
+				return new RowDefinitionCollection(lengths.Select(length => new RowDefinition {Height = (GridLength)converter.ConvertFromInvariantString(length) }).ToArray());
```
Comment (eerhardt) [inline src/Controls/src/Core/RowDefinitionCollectionTypeConverter.cs:30]: I would suggest not to use Linq Select and ToArray here and just use the foreach as before. The performance overhead isn't worth it. We spend a decent amount of time removing Linq usages in dotnet/runtime.

File: src/Controls/src/Core/RowDefinition.cs [hunk 12]
```diff
@@ -6,13 +6,17 @@ namespace Microsoft.Maui.Controls
...
+		{
+			SetValue(HeightProperty, height);
```
Comment (eerhardt) [inline src/Controls/src/Core/RowDefinition.cs:19]: (nit)

```suggestion
			Height = height;
```

File: src/Controls/src/Build.Tasks/CompiledConverters/ColumnDefinitionCollectionTypeConverter.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+			if (!string.IsNullOrWhiteSpace(value))
+			{
+				value = value.Trim();
+				var gridlengthconverter = new GridLengthTypeConverter();
+				var parts = value.Split(',');
+
+				yield return Create(Ldc_I4, parts.Length);
+				yield return Create(Newarr, module.ImportReference(("Microsoft.Maui.Controls", "Microsoft.Maui.Controls", "ColumnDefinition")));
+				for(var i =0;i< parts.Length;i++){
+				        yield return Create(Dup);
```
Comment (eerhardt) [inline src/Controls/src/Build.Tasks/CompiledConverters/ColumnDefinitionCollectionTypeConverter.cs:25]: The whitespace in this file is off. `CTRL+K+D` is your friend 😉

File: src/Controls/src/Build.Tasks/CompiledConverters/ColumnDefinitionCollectionTypeConverter.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+					yield return Create(Ldc_I4, i);
+					foreach (var instruction in gridlengthconverter.ConvertFromString(parts[i], context, node))
+						yield return instruction;
+					yield return Create(Newobj, module.ImportCtorReference(
```
Comment (eerhardt) [inline src/Controls/src/Build.Tasks/CompiledConverters/ColumnDefinitionCollectionTypeConverter.cs:29]: Does calling `ImportCtorReference` in a loop affect performance at all? I assume we would normally do this outside the loop.

Comment (StephaneDelcroix) [reply src/Controls/src/Build.Tasks/CompiledConverters/ColumnDefinitionCollectionTypeConverter.cs:29]: that's fine, as 
1/ it's compiler time
2/ Import*Reference calls and return values are cached

File: src/Controls/src/SourceGen/Controls.SourceGen.csproj [hunk 15]
```diff
@@ -30,7 +30,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.0.0-4.final" PrivateAssets="All" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.2.0-1.final" PrivateAssets="All" />
```
[the remark this answers is not in the document]
Comment (StephaneDelcroix) [reply src/Controls/src/SourceGen/Controls.SourceGen.csproj]: reverting, no reason to bump at this time

File: src/Controls/src/Core/RowDefinition.cs (hunk 12 above)
Comment (StephaneDelcroix) [reply src/Controls/src/Core/RowDefinition.cs:19]: the actual setting of value is done by SetValue, even if you set the property. This is one invocation saved. See this as my contribution to the global warming problem

File: src/Core/src/Primitives/GridLength.cs (hunk 5 above)
Comment (eerhardt) [reply src/Core/src/Primitives/GridLength.cs:12]: If we didn't want to make a breaking change here, we could accomplish the same thing by using a property with an initializer:

```C#
public static GridLength Auto { get; } = new GridLength(1, GridUnitType.Auto);
```

**Review by eerhardt — APPROVED** [review]
LGTM.

You may want to consider https://github.com/dotnet/maui/pull/5489#discussion_r833401903. Typically we use properties instead of fields in public APIs.

**Pull request merged**